Cached results support

pierresouchay opened the pull request:
Sync with master of rickfast/consul-client + added support for CACHED results in Consistency

PR on upstream: https://github.com/rickfast/consul-client/pull/405

yfouquet committed:
Fix flaky tests (#320)

* Fix flaky ServiceCatalogCacheTest

The test ServiceCatalogCacheTest#testWatchService is flaky because:
 - the cache#start method is asynchronous
 - the call to cache#register is done directly after the cache#start
 - if the service is registered before the cache is started, the test fails

Calling cache#awaitInitialized before registering new services fix the test.
This is the same thing which is done is all other cache tests.

* Fix flaky SnapshotClientTest

Use temporary file and delete it.

yfouquet committed:
Upgrade jackson dependency to version 2.9.5 (#319)

Commit:
1.1.2

yfouquet committed:
Fix Snapshot test flakyness (#322)

The service can be not registered yet when chechking; adding a small pause should fix it.
The latch is now passed to the callback directly.

yfouquet committed:
Fix Snapshot test flakyness (2) (#323)

This test is flaky, fix it.

yfouquet committed:
Upgrade consul version in test to 1.0.7 (#324)

This new version add several features and bug fixes including service meta-data as stated in https://github.com/rickfast/consul-client/issues/321
For more details: https://github.com/hashicorp/consul/blob/master/CHANGELOG.md#107-april-13-2018

yfouquet committed:
Add support for service meta data (#325)

We should be able to:
 - register a service in an agent with meta data
 - register a service in the catalog with meta data
 - retrieve the meta data of a service (catalog, health, cache)

Commit:
1.2.0

Commit:
Support EnableTagOverride during service registration (#331)

* Support EnableTagOverride during service registration

* Back out adding extra line

yfouquet committed:
Enable custom log function in case of cache refresh error (#334)

We need to log sometimes as error, sometimes as warning.
Also, we would like to not log the same error every time (wait a few minutes in between).
This is specific to our use case.

Hence, we here allow setting a custom consumer to log on cache refresh error.
The behavior of caches for those who don't need this feature will remain unchanged.

yfouquet committed:
Add more details in cache logs (#336)

When several caches are used by the same app, it is hard to know which cache generated a log.
Add cache descriptor in the cache logs.

yfouquet committed:
Display logs in test (#335)

When playing tests, we might want to see the logs.
However, logs are not available as no implementation of slf4j is present.
Add slf4j implementaion for test only.

Example of output:
10:23:15.307 [OkHttp http://localhost:8500/...] DEBUG com.orbitz.consul.cache.ConsulCache - Consul cache updated (index=672), request duration: 274 ms

yfouquet committed:
Allow using randomized back-off delay (#333)

Use uniformly distributed back-off delay within min/max delay on retry.

See issue #332

yfouquet committed:
Improve cache event callback (#338)

* Use better cache descriptor for monitoring

The cache descriptor is a simple string.
Split it in endpoint and key for better monitoring.

* Add cache descriptor in event callbacks for cache

When receiving a notification like onCacheStart(), it would be better to know which cache has started.
Hence, send the cache descriptor with each cache event callback.

* Add comment on cache descriptor

yfouquet committed:
Split unit tests and integration tests (#337)

* Split unit tests and integration tests

Integration tests requires a consul agent to be running with correct configuration.
When developing a feature, it would be great to be able to launch unit tests first.
Hence, we here purpose to:
 - have unit test classes in src/test with Test suffix
 - have integration test classes in src/itest with ITest suffix

* Fix agent test

Commit:
1.2.1

yfouquet committed:
Enable setting args when registering service check (#341)

* Enable setting args when registering service check

* Remove deprecated register method

As we cleaned up some registration method, let's remove deprecated ones.

yfouquet committed:
[Fix] Enable setting several datacenters in prepared query failover (#344)

As stated in #274, a string is not a valid input for the datacenters list.
Use a list of strings instead.
See https://www.consul.io/api/query.html#failover

yfouquet committed:
[Fix] Allows for ACL in getKeys() method (#345)

As stated in #253, we cannot add token in getKeys() methods.
However, it is possible in Consul (see https://www.consul.io/api/kv.html#read-key)

Add a way to do so.

yfouquet committed:
Update documentation (#348)

Added all code samples in a test file to ensure they remain relevant.
As we cannot include file sections yet, we must copy samples from the test file.

yfouquet committed:
Add support for non available lan address (#347)

Lan address was added in Consul 0.7.
Lower versions of Consul require this field to be optional.
Note that it does not have any impact on recent Consul versions.

Commit:
1.2.2

Commit:
Merge branch 'master' of https://github.com/rickfast/consul-client

Commit:
1.2.3

Benjamin-L-Dev committed:
Fixed annotations on API for AgentClient's join request (#349)

yfouquet committed:
[Fix] Agent/force-leave does not work (#353)

This was due to the fact that the `node` parameter was missing.
The endpoint agent/force-leave does not exist and Consul respond with 301.
The request was then redirected to agent/force-leave/ by OkHttp client.
This redirection is always done with GET and Consul rejects it once more.

The API documentation of Consul HTTP API (hashicorp) is out-of-date.
To see the actual endpoint see:
 - https://github.com/hashicorp/consul/blob/master/agent/http_oss.go#L25
 - https://github.com/hashicorp/consul/blob/master/agent/agent_endpoint.go#L323
 - https://github.com/hashicorp/consul/blob/master/agent/agent.go#L1442

Commit:
Add ServiceWeights property to CatalogService (#357)

* Add Weights property to Service and CatalogService

Consul recentlty added support for service weights. (see
https://github.com/hashicorp/consul/issues/4198).
This change adds the new field to the Service and CatalogService
classes.

* Use latest (1.2.3) consul version for travis

Commit:
1.2.5

heanssgen-troy committed:
Add Multiple Host Failover Support To Consul Client (#354)

* Added multiple hostname support. This is implemented using the underlying okhttp3 Interceptor, which allows us to replicate requests across multiple hosts if they fail

* Fixed indentation to match spaces instead of tabs

* Refactored the new failover interceptor out of the main consul client
and into a package for clarity. Also added a host selection strategy so
that others can extend a way of selecting additional failover hosts from
a set (if needed).

* Removed bad debugging message

Commit:
Merge branch 'master' of https://github.com/rickfast/consul-client

Commit:
1.2.6

Commit:
Add minDelayOnEmptyResult in cache config (#360)

In cache, when an empty response is received, wait at least
minDelayOnEmptyResult before running the next request. If
minDelayBetweenRequest is greater than minDelayOnEmptyResult,
wait minDelayOnEmptyResult instead.

heanssgen-troy committed:
A 400 series error should not blacklist the consul instance that is (#362)

responding. This is because a 404 or 403 error is actually completely
valid.

This does remove some potential functionality (such as being able to do
cross-cluster failover), however that functionality is not exactly
supported anyways since it introduces the potential to set a key in one
cluster and then fail to watch it in another.

Commit:
1.2.7

rickfast committed:
Removing Bintray

rehevkor5 committed:
Listener exceptions no longer impact other listeners (#364)

- Exceptions thrown by listeners are now caught so that the remaining
listeners will still be called. Such exceptions are logged at
WARN level.
- runCallback() must execute in the executorService thread, otherwise
if the delay is set to zero, StackOverflow can occur.
- Added fail-fast null checks for ConsulCache constructor params.
Previously, NPE would have been thrown later on.
- Added name format for ConsulCache's scheduled executor, to
differentiate that thread from others in stack dumps, logs, etc.
- Fixed StackOverflowException when using a CallbackConsumer which
does not do its work in a separate thread, and when the minimum duration
between requests config value is set to zero.
- Added param to docker command in README to match travis config, so
that integration tests work when running locally.

Commit:
1.3.0

rickfast committed:
Security updates for Retrofit and Jackson (#365)

pellepelster committed:
add initial implementation for the new consul policy and token api (#366)

* add initial implementation for the new consul policy api

* split up environment setup so it can be re-used in travis ci build

* move policy handling into acl client to better reflect consul api layout

* add token removal and listing functionality

* update documentation to reflect enviroment setup changes

Commit:
Shutdown event executor (#367)

* DC is optional in catalog entity deregistration https://www.consul.io/api/catalog.html#deregister-entity

* QueryOptions in catalog registration and deregistration - in order to pass ACL token

* add possibility to shutdown event-executor threads

* problem with propagating events should not break stopping behaviour

Commit:
1.3.1

Commit:
add the possibility to customize a ConnectionPool for OkHttp client (#371)

default behaviour of a ConnectionPool it's to cleanup connection
threads after 5 minutes, that is not suitable for all applications

sunchuanleihit committed:
bugfix consulResponse method throw NumberFormatException interupt callback (#370)

Commit:
1.3.2

Commit:
Upgrade dependencies (#377)

Commit:
1.3.2

Commit:
add `GET agent/service/:service_id` implementation (#382)

Commit:
1.3.4

Commit:
Add X-Consul-Token support (#384)

Commit:
1.3.5

Commit:
Fix for #381 (#385)

Commit:
Fix for #358 (#386)

Commit:
1.3.6

Commit:
upgrade dependencies

studouglas-scx committed:
include queryOptions in listEvents API calls (#390) (#391)

Commit:
jackson 2.9.9.2 does not exist

alespuh committed:
Let ConsulCache clients chose what ScheduledExecutorService should be used by ConsulCache (#389)

Commit:
1.3.7

alespuh committed:
If ScheduledExecutorService was provided externally then don't shutdown it during stopping a cache (#396)

Commit:
1.3.8

Commit:
rev jackson-databind to remediate security vuln

cbarbara-okta committed:
Support for https only environments (#400)

Added the `withHttps` method to the Consul builder class that can set your url scheme to be "https".
Added a debug log message to `ConsulFailoverInterceptor` since it is the only place to find the full url and server where the failed request happened.

cbarbara-okta committed:
Bugfix into `BlacklistingConsulFailoverStrategy` so that it can return an empty Optional value when all consul hosts have been blacklisted.  Previously it would throw a `NoSuchElementException` instead of returning an empty value.  This allows for the `ConsulFailoverInterceptor` to correctly throw a `ConsulException` in these cases. (#399)

rickfast committed:
Update .travis.yml

pierresouchay committed:
Added support for cached results consistency

Many endpoints including health now support cached results, aka
the local Consul agent will sync with the servers to get updates.

This includes support for this mode, especially useful when several
applications are using the same services.

See https://www.consul.io/api/features/caching.html

ShimmerGlass commented:
For cached query it is important to specify the Cache-Control header to limit how old the results can be. ex: `Cache-Control: max-age=30 stale-if-error=259200`.
Without this there is no limit to how old the result can be if something goes wrong. cf: https://www.consul.io/api/features/caching.html

pierresouchay committed:
Allow to specify cache control as well as getting information about cache HIT/MISS

While the first patch was working, it was a bit insufficient as it did not allow
to handle precise cache Control as suggested by @Aestek

I took the path to change the enum ConsistencyMode into a class emulating enum's
methods in order to keep backwards compatibility and avoiding breaking clients.

This patch is a bit intrusive since I have to dynamically change Header Cache-Control
during requests and current static ways to inject headers using injectors was not
sufficient. The mechanism might be extended later to dynamically inject other kind of
headers if necessary.

I also added a way to get Cache result information when available in Consul Response.

Cache has been implemented only for endpoints supporting it.

ShimmerGlass commented:
LGTM

pierresouchay committed:
Re-implemented toString() in ConsistencyMode to have same behaviour as enums for existing values

pierresouchay commented:
Merged on upstream, already integrated in our fork
